### PR TITLE
Update ImageWidget to support more stretch settings, and enable screensaver on kobo

### PR DESCRIPTION
**Pull request by Hzj-jie**
This issue has been discussed @ #1968 .

Also, current ImageWidget does not provide enough stretch strategies to better present a screen saver.
I have tried several combinations, and finally decide to use overflow + centering, considering users may use screensaver to hide what they are reading, so stretch the image to cover entire screen but respect original aspect ratio should be a good choice.

### CODE REVIEW

**Comment by NiLuJe** [comment]
FWIW, for the Kindle ScreenSaver hack, my favorite choice is keeping the aspect ratio, and letterboxing w/ black borders to cover the screen.

File: reader.lua [hunk 1]
```diff
@@ -97,7 +97,12 @@ local UIManager = require("ui/uimanager")
 local Device = require("device")
 local Font = require("ui/font")
 
--- read some global reader setting here:
+-- change some global default values according to the device types
+if Device:isKobo() then
+    KOBO_SCREEN_SAVER = "/mnt/onboard/screensaver"
```
Comment (houqp) [inline reader.lua]: Can you put this into https://github.com/koreader/koreader/blob/master/frontend/device/kobo/device.lua#L81?

**Comment by houqp** [comment]
+1 on @NiLuJe 's idea. @Hzj-jie if you think that's too much work to implement, we can merge this one and leave that as a new feature request ticket.

File: reader.lua (hunk 1 above)
Comment (houqp) [inline reader.lua]: Actually, wouldn't this override the setting in defaults.lua?

Comment (Hzj-jie) [inline reader.lua]: Yes, it overrides, which is expected.

**Comment by Hzj-jie** [comment]
Let me try, I believe an AlphaContainer should fulfill the requirement.

**Comment by houqp** [comment]
How should a user customize the screensaver directory after this is merged?

**Comment by Hzj-jie** [comment]
Currently there is not a setting to change the directory, so it's device dependent. I do not have a pocketbook, it may have a different directory to mount its internal storage.

**Comment by houqp** [comment]
If we are planning to not make this configurable, might as well just remove the config from defaults.lua. pocketbook users should not be using this config anyway since it starts with `KOBO`.

**Comment by Hzj-jie** [comment]
I prefer to remove the KOBO_ prefix from the names of these defaults. Since both kobo and pocketbook use these configurations.

**Comment by KenMaltby** [comment]
So are you removing the ability for the user to add their own image file?  Or just removing the ability to set the image file/directory in defaults.lua?   Would "/mnt/onboard/screensaver"  not be where the image is placed?

**Comment by Hzj-jie** [comment]
Neither,
1. There was not a default setting in defaults.lua.
2. /mnt/onboard/screensaver is the new default setting for kobo, but since I do not have a pocketbook, I have no idea which mount point users can write on pocketbook.

**Comment by KenMaltby** [comment]
What about line 181: KOBO_SCREEN_SAVER = ""             -- image or directory with pictures or "-" 

(I don't know why it is under "-- Light parameter for Kobo", though?)

**Comment by houqp** [comment]
@Hzj-jie 

But by overriding KOBO_SCREEN_SAVER in kobo:init, kobo users won't be able to set the directory in defaults.lua anymore. Am I missing anything here?

@KenMaltby 

> I don't see it as beneficial for KOReader to be adding an image building/manipulation function, the generation of images is something for which there are a number of good programs. GIMP2, and Photoshop come to mind. If you want a better image, use one of the available programs to create it.

Yes, ideally, people should use professional image processing tools to fit the image on to the device screen. In that case, this change will display the image fine. The auto scaling is just for people won't don't have the time and expertise to do the processing bit.

**Comment by houqp** [comment]
> What about line 181: KOBO_SCREEN_SAVER = "" -- image or directory with pictures or "-"
> 
> (I don't know why it is under "-- Light parameter for Kobo", though?)

It should not be, someone added this config without adding the comment for a new section. Anyway, I would prefer all these are set in reader setting instead of global defaults :(

**Comment by Hzj-jie** [comment]
Oh, so do you expect user to set defaults.lua file? If so, it's my fault. I thought all the configurations are controlled by UI.

**Comment by houqp** [comment]
> Oh, so do you expect users to change defaults.lua file? If so, it's my fault. I thought all the configurations were controlled by UI, or in reader.settings.lua, since it's not included in our package.

Yes, defaults.lua was added a long time ago as a quick way for user to customize some of the reader behaviors. Because we ship with defaults.lua, `defaults.persistent.lua` was introduced so people's changes won't be overwritten by new updates.

This is a hack and we should stop using it. As I said before, I also prefer everything to be controlled in reader.settings.lua.

> And if we expect users to change defaults.lua or add persistent/defaults.lua, it would be better to set the default value in defaults.lua instead. Then persistent/defaults.lua can override this setting, if an advanced user want to.

Sounds good to me, we can set the default there.

**Comment by Frenzie** [comment]
@Hzj-jie Could it be an idea to store the generated image somewhere appropriate?

@KenMaltby I don't see basic aspect ratio retention as adding an image manipulation program. The fact that ebook covers get stretched in the weirdest ways is just ugly. Or was that fixed already? When you get right down to it I guess I really don't care about the screensaver. =P Like I've said elsewhere, I'd prefer to sleep with the book "open", like in EPUBs without a cover.

**Comment by KenMaltby** [comment]
@Frenzie; We already have a Reader Menu "Screensaver" option: "Display proportional cover image in screensaver" , so it might be extended to a user selected image.  

I have suggested having a "Long Press" on a displayed image offer the option to make it the suspend/screensaver image.  I suspect having a defined place and name for the image, to be used, would be required.  I have no problem with that, I just wouldn't think it a good idea bar the user from being able to substitute one of their own. 

I also still think being able to have a static display of the Book Status page, as the screensaver, would be a good idea (and I could care less about reading statistics).  [BTW, it is easy to comment out the "show" command, for the suspend message box, and have the current page remain up]

**Comment by Frenzie** [comment]
@KenMaltby Ah, right. So the problem this PR is trying to solve is that *if* you want a custom screensaver, the aspect ratio isn't kept? The whole aspect ratio calculation seems to be added afresh rather than reused or refactored from elsewhere, which is why I assumed there was none at present.

**Comment by KenMaltby** [comment]
Just for the record, the UI does include setting the default.lua options, it is what the "Advanced Options" function accomplishes.

**Comment by Frenzie** [comment]
@KenMaltby I thought this PR was about keeping the AR of one's own custom image?

**Comment by KenMaltby** [comment]
> @KenMaltby I thought this PR was about keeping the AR of one's own custom image?

Just going by the title, it seemed to offer more stretching options but also impacted the ability for the user to provide their own custom image.  Or at least a change in how that would be accomplished. Not sure I follow what is in there now.

**Comment by houqp** [comment]
Based on our previous discussion, I think @Hzj-jie agreed to make `KOBO_SCREEN_SAVER` configurable. @KenMaltby , do you have any other concern?

**Comment by KenMaltby** [comment]
No concerns with this, it might be a practical step in the direction of an overall improvement in rendering and displaying images in KOReader.

**Comment by Hzj-jie** [comment]
Based on our discussion, should we set a rule, for constants users cannot set via UI should be put in defaults.lua, and always be able to overwrite by /defaults.persistent.lua. Other constants, settings, should be in reader.settings.lua?

**Comment by houqp** [comment]
> Based on our discussion, should we set a rule, for constants users cannot set via UI should be put in defaults.lua, and always be able to overwrite by /defaults.persistent.lua. Other constants, settings, should be in reader.settings.lua?

This should be the rule. Although I would prefer we stop adding more global constatns in defaults.lua and move everything over to reader.settings.lua from now on.

**Comment by Hzj-jie** [comment]
I think default.lua is still a good place for settings for advanced users, they can set their preferences in default.persistent.lua for some features cannot be configured with UI. Separating 'regular user settable features' and 'advanced users only features' looks like a good choice.

**Comment by houqp** [comment]
If we really want a advanced setting section, we can also just add one more section in reader.settings.lua ;)

Looking at the current defaults.lua, a lot of the settings should really be user settable features. I think we have been abusing this file for a long time because it's the easiest way to introduce a user configurable variable. However, global variables make the code harder to maintain and adds performance penalty when running under LuaJIT.

There are some rare cases where reader.settings.lua will not work, then we can use defaults.lua. Otherwise, it's better to stick with reader.settings.lua.

**Comment by Hzj-jie** [comment]
Reasonable, I will modify the change to move the setting into reader.settings.lua.

**Comment by KenMaltby** [comment]
So, should there be a Reader Menu "Advanced settings" item, like the File Manager one for defaults.lua - except for "reader.settings.lua" ?

**Comment by Hzj-jie** [comment]
Yes, you can set 'screensaver_folder' to any folder readable & writable.

**Comment by houqp** [comment]
We should probably get a proper setting interface rather than adding everything into a main menu ;) Let's take stab at it when @chrox finishes up koreader home.

**Comment by KenMaltby** [comment]
What happened to the "Screensaver" menu items?  Can we no longer turn on and off use of the book's cover?  Also, there was an option to use or not use a "proportional" display of the cover image.

**Comment by houqp** [comment]
Sorry, i meant to say `settings.reader.lua`. Do you remember where the screensaver menu item was? Is it in the advanced setting?

**Comment by KenMaltby** [comment]
Not in the "Advanced settings", it was in the upper Reader menu.  There was a KOBO_SCREEN_SAVER_LAST_BOOK = true -- get screensaver from last book if possible, that was in defaults.lua but I don't know if there is anything anymore.

**Comment by houqp** [comment]
Oh, yeah, we removed it from defaults.lua. I will add a menu setting for it.

**Pull request merged**